File: 021.MergeTwoSortedLists/MarkCSIE/MergeTwoSortedLists/src/MergeTwoSortedLists.java

class ListNode {

    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

public class MergeTwoSortedLists {

    public static void main(String[] args) {
        ListNode testCase1 = new ListNode(1);
        testCase1.next = new ListNode(1);
        testCase1.next.next = new ListNode(1);
        testCase1.next.next.next = new ListNode(2);
        testCase1.next.next.next.next = new ListNode(3);

        ListNode testCase2 = new ListNode(1);
        testCase2.next = new ListNode(1);
        testCase2.next.next = new ListNode(1);
        testCase2.next.next.next = new ListNode(2);
        testCase2.next.next.next.next = new ListNode(3);
        
        MergeTwoSortedLists test = new MergeTwoSortedLists();
        ListNode testAnswer1 = test.mergeTwoLists(testCase1, testCase2);
        while (testAnswer1 != null) {
            System.out.print(testAnswer1.val);
            System.out.print(" ");
            testAnswer1 = testAnswer1.next;
        }
        System.out.println();
    }

    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        } else if (l2 == null) {
            return l1;
        }

        ListNode head = null;
        ListNode currNode = head;
        while (true) {
            if (l1 == null) {
                currNode.next = l2;
                break;
            }
            if (l2 == null) {
                currNode.next = l1;
                break;
            }

            ListNode targetNode;
            if (l1.val <= l2.val) {
                targetNode = l1;
                l1 = l1.next;
            } else {
                targetNode = l2;
                l2 = l2.next;
            }

            if (currNode == null) {
                head = targetNode;
                currNode = targetNode;
            } else {
                currNode.next = targetNode;
                currNode = currNode.next;
            }
        }
        return head;
    }
}

File: 021.MergeTwoSortedLists/Ddfish/MergeTwoSortedLists.swift

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

class Solution {
    
    func mergeTwoLists(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        
        guard let l1 = l1 else {
            return l2
        }
        guard let l2 = l2 else {
            return l1
        }
        
        var answer : ListNode?
        
        if l1.val < l2.val {
            answer = ListNode(l1.val)
            answer?.next = mergeTwoLists(l1.next, l2)
        }else{
            answer = ListNode(l2.val)
            answer?.next = mergeTwoLists(l1, l2.next)
        }
        
        return answer
    }
    
}
